src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -20,6 +20,13 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncComplet
 {
     internal sealed partial class ItemManager : IAsyncCompletionItemManager2
     {
+        /// <summary>
+        /// The threshold for us to consider including (potentially large amount of) expanded items in completion list.
+        /// Showing a large amount of expanded items to user would introduce noise and render the list too long to browse. 
+        /// Not processing those expanded items also has perf benefit (e.g. matching and highlighting could be expensive)
+        /// </summary>
+        internal const int MaximumFilterTextLengthToExcludeExpandedItems = 3;
```

Comment:
IMO this should be 2. As a general goal, it should not require more that 3 correctly-typed characters to produce the completion item of interest. The current value of 3 forces it to be 4.

Reply:
I agree with sam that this should go down to 2 for the general case.  But i would go down to '1' for the 'all caps' case as well.

Note: it's unclear why there's any perf concern in this space as:

1. this should all be async.
2. this should all be done in teh bg.
3. the completion list api allows us to add items in batches.  so any slow set (like the unimported set) shouldn't impact the perf of the fast set.

Reply:
> the completion list api allows us to add items in batches. so any slow set (like the unimported set) shouldn't impact the perf of the fast set.

Correct behavior of completion (specifically regarding selection and commit characters) cannot be time-dependent, which means all batches must be considered in response to a completion trigger.

Reply:
> I agree with sam that this should go down to 2 for the general case. But i would go down to '1' for the 'all caps' case as well.

Of course, that makes sense.

>  it's unclear why there's any perf concern in this space 

Yes, it's async and in background. But while we can add items in batches, the issue reported by Editor folks is after the "slow set" is available. Sorting/matching/highlighting a larger number of items takes longer, it adds ~40ms of delay for showing refreshed list for every keystroke in 90th percentile (compare to 7ms before).

Reply:
can we not do the sorting for them?  aren't we already doing pattern matching on our end to even add the imported items?

Reply:
> All the operations I listed is done by us, not editor. 

Why do they care then?  the fast set will still be fast.  The slow set will take a little more time... but it's only for the additional values we're adding, which only provide benefit to hte user.  In other words, we're making ourselves faster here by taking away things user may want.   It does not help any users (since hte completion list is still fast for the normal items), and it hurts the user that wants these items.

I'm actually not seeing what problem we're trying to solve here.

>  it adds ~40ms of delay for showing refreshed list for every keystroke in 90th percentile (compare to 7ms before).

But this delay is only for that other batch of items right?  So why is that an issue...?